aptos-move/framework/aptos-framework/sources/delegation_pool.move

```diff
@@ -0,0 +1,1719 @@
+                    pending_inactive - commission_pending_inactive
+                );
+                // if withdrawal's lockup cycle was ended ONLY on stake pool then its stake is inactive
+                if (lockup_cycle_ended) { (pending_inactive, 0) } else { (0, pending_inactive) }
+            }
+        } else { (0, 0) };
+
+        // should also include commission rewards in case of the operator account
+        // operator rewards are actually used to buy shares which is introducing
+        // some imprecision (received stake would be slightly less)
+        // but adding rewards onto the existing stake is still a good approximation
+        if (delegator_address == stake::get_operator(pool_address)) {
+            active = active + commission_active;
+            if (lockup_cycle_ended) {
+                inactive = inactive + commission_pending_inactive;
+            } else {
+                pending_inactive = pending_inactive + commission_pending_inactive;
+            }
+        };
+        (active, inactive, pending_inactive)
+    }
+
+    #[view]
+    /// Return the unique observed lockup cycle where delegator `delegator_address` may have
+    /// unlocking (or already unlocked) stake to be withdrawn from delegation pool at `pool_address`.
+    public fun get_pending_withdrawal(pool_address: address, delegator_address: address): (bool, u64) acquires DelegationPool {
+        assert_delegation_pool_exists(pool_address);
+        pending_withdrawal_exists(borrow_global<DelegationPool>(pool_address), delegator_address)
+    }
+
+    /// Return the unique observed lockup cycle where delegator `delegator_address` may have
+    /// unlocking (or already unlocked) stake to be withdrawn from delegation pool `pool`.
+    /// A bool is returned to signal if a pending withdrawal exists at all.
+    fun pending_withdrawal_exists(pool: &DelegationPool, delegator_address: address): (bool, u64) {
+        if (table::contains(&pool.pending_withdrawals, delegator_address)) {
+            (true, *table::borrow(&pool.pending_withdrawals, delegator_address))
+        } else {
+            (false, 0)
+        }
+    }
+
+    /// Return a mutable reference to the shares pool of `pending_inactive` stake on the
+    /// delegation pool, always the last item in `inactive_shares`.
+    fun pending_inactive_shares_pool(pool: &mut DelegationPool): &mut pool_u64::Pool {
+        let observed_lockup_cycle = pool.observed_lockup_cycle;
+        table::borrow_mut(&mut pool.inactive_shares, observed_lockup_cycle)
+    }
+
+    /// Buy shares into active pool on behalf of delegator `shareholder` who
+    /// delegated `coins_amount` stake or reactivated(redeemed) it from pending-inactive pool.
+    fun buy_in_active_shares(
+        pool: &mut DelegationPool,
+        shareholder: address,
+        coins_amount: u64,
+    ): u64 {
+        if (coins_amount == 0) return 0;
```

Comment:
nit but less code = better IMO

I don't think this is needed as pool_u64::buy_in already does this. This can also help deleting this extra buy_in_active_shares function and just inline the pool_u64::buy_in call